ffs/scheduler/internal/sjstore/sjstore.go

```diff
@@ -360,111 +369,169 @@ func (s *Store) GetStartedDeals(iid ffs.APIID, c cid.Cid) ([]cid.Cid, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	var iids []ffs.APIID
-	if iid == ffs.EmptyInstanceID {
-		for iid := range s.queuedJobs {
-			iids = append(iids, iid)
-		}
-	} else {
-		iids = append(iids, iid)
-		ensureJobsSliceMap(s.queuedJobs, iid)
-	}
-
-	var res []ffs.StorageJob
-	for _, iid := range iids {
-		cidsList := cids
-		if len(cidsList) == 0 {
-			for cid := range s.queuedJobs[iid] {
-				cidsList = append(cidsList, cid)
+	byTime := func(ascending bool) func(a, b query.Entry) int {
+		extractTime := func(key string) (int64, error) {
+			parts := strings.Split(key, "/")
+			if len(parts) != 5 {
+				return 0, fmt.Errorf("expected 5 key parts but got %v", len(parts))
 			}
+			return strconv.ParseInt(parts[4], 10, 64)
 		}
-		for _, cid := range cidsList {
-			jobs := s.queuedJobs[iid][cid]
-			for _, job := range jobs {
-				res = append(res, *job)
+		return func(a, b query.Entry) int {
+			l := a
+			r := b
+			if ascending {
+				l = b
+				r = a
+			}
+			lTime, err := extractTime(l.Key)
+			if err != nil {
+				log.Errorf("extracting time from key a: %v", err)
+				return 0
+			}
+			rTime, err := extractTime(r.Key)
+			if err != nil {
+				log.Errorf("extracting time from key b: %v", err)
+				return 0
+			}
+			if lTime > rTime {
+				return -1
+			} else if rTime > lTime {
+				return 1
+			} else {
+				return 0
 			}
 		}
 	}
 
-	// Return these newest first.
-	sort.Slice(res, func(a, b int) bool {
-		return res[a].CreatedAt > res[b].CreatedAt
-	})
-
-	return res
-}
-
-// ExecutingJobs returns executing jobs for the specified instance id and cids.
-// If the instance id is ffs.EmptyInstanceID, data for all instances is returned.
-// If no cids are provided, data for all data cids is returned.
-func (s *Store) ExecutingJobs(iid ffs.APIID, cids ...cid.Cid) []ffs.StorageJob {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
-	return mappedJobs(s.executingJobs, iid, cids...)
-}
-
-// LatestFinalJobs returns the most recent finished jobs for the specified instance id and cids.
-// If the instance id is ffs.EmptyInstanceID, data for all instances is returned.
-// If no cids are provided, data for all data cids is returned.
-func (s *Store) LatestFinalJobs(iid ffs.APIID, cids ...cid.Cid) []ffs.StorageJob {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+	// by default, query all.
+	prefix := dsBaseCid.String()
+	if config.APIIDFilter != ffs.EmptyInstanceID && config.CidFilter.Defined() {
+		// use apiid/cid
+		prefix = prefixAPIIDAndCid(config.APIIDFilter, config.CidFilter)
+	} else if config.APIIDFilter != ffs.EmptyInstanceID && !config.CidFilter.Defined() {
+		// use appid
+		prefix = prefixAPIID(config.APIIDFilter)
+	} else if config.APIIDFilter == ffs.EmptyInstanceID && config.CidFilter.Defined() {
+		// use cid
+		prefix = prefixCid(config.CidFilter)
+	}
 
-	return mappedJobs(s.lastFinalJobs, iid, cids...)
-}
+	q := query.Query{
+		Prefix: prefix,
+		Orders: []query.Order{query.OrderByFunction(byTime(config.Ascending))},
```

Comment:
Here's an important point that I mentioned it some time ago when discussing about how Query implementations work.

Whenever you set some `Orders`, what implementations will do is fetch all results from the datastore in memory, and then apply the sorting. So this means that all queries will load all things in memory, sort them, and then make them available in the for loop.

For some cases in the above if, the `Orders` could be avoided since by default keys are iterated in lexicographic order since you also added createdAt at the end of the key. At least, for Ascending ordering seems that `Orders` can be avoided, but descending ones we are out of luck (really unfortunate).

Does this make sense to you?

Reply:
A couple thoughts there.

- This is a query of only the index data, so yes, to order it has to load all the data, but at least it's just simple index data and not the actual data.
- I wanted to make descending ordering the default since I think that is the most common way to look at the results. "show me what is going on recently" way of thinking. So to support that, no default lexicographic order would work unfortunately. Plus:
- I don' think these index keys actually sort ascending lexicographically by time because the api id and cid comes before the timestamp in the key like: `apiid/<apiid>/<cid>/<timestamp>. Right?

Reply:
@jsign, I'm still considering using in memory maps instead of go-datstore indexes, but in general do you view either of those to be a workable solution or are you viewing the ordering issue or the way these indexes work to be a no-go thing that needs to change more fundamentally?

Reply:
- 👍🏼 
- 👍🏼 
- But in some of your queries you're doing `prefix = prefixAPIIDAndCid(config.APIIDFilter, config.CidFilter)`, wouldn't those keys lexicographic order should only be determined by the last part of the key? (timestamp) Since APIID and Cid are fixed. Recall that I mentioned that in "some case" you could at least leverage this. Missing something?

Reply:
> @jsign, I'm still considering using in memory maps instead of go-datstore indexes, but in general do you view either of those to be a workable solution or are you viewing the ordering issue or the way these indexes work to be a no-go thing that needs to change more fundamentally?

I think your implemented idea is close to what I'd do with go-datastore. So I'm not much concerned about the design or implementation; the only ugly thing is that you'd need migration to re-create those indexes now. And also, I was most concerned about some code complexity and not sure about how bad would be to do this with go-datastore (`Order` field, etc).

So to be clear, I think the "cons" aren't really in your implementation or design (I think that's OK). But mostly to some unavoidable complexity or question-marks around the cost of all this (being in complexity or perf) caused by go-datastore API.

Reply:
Re: Query prefixing and lexicographic order... Right, it would only work for a subset of queries that specify both api id and cid. If you specified only api id or only cid, you'd have to use `Order` still.

Thanks for clarifying what you think the "cons" are and what that all means.

Reply:
> If you specified only api id or only cid or neither, you'd have to use Order still
Well, not quite true... I could also add two more indexes: `/apiid-only/<api-id>/timestamp` and `/cid-only/<cid>/timestamp`, but I was trying to find some optimal situation where the number of indexes wasn't crazy and there are still tools available to query the data in the required ways with acceptable performance. Plus the whole descending order by default thing removes value from making all possible indexes that will support ascending lexicographic order.

Reply:
Yeah, all things can be solved with more indexes.
My point in my original thread comment was that there was a special case of your if-cases that can avoid the `Order`.

I'm totally against adding complexity in general (in this case, adding more indices). 
I just mentioned a special-case opportunity to make it faster without adding new indexes or wild stuff.

Reply:
Gotcha, thanks.